### DISCUSSION TITLE
Unit testing OpenAI dotnet classes - internal + sealed

### DISCUSSION THREAD

**Discussion by gjz22** [microsoft/semantic-kernel#9388] [Q&A] [answered]
We use OpenAI chat completion service, but OpenAIChatMessageContent is sealed and has an internal method constructor. Are there any ways people are unit testing classes that use this object?

**Comment by sophialagerkranspandey** [comment, marked as answer]
We recommend checking out details on Unit Testing here: https://devblogs.microsoft.com/semantic-kernel/unit-testing-with-semantic-kernel/

**Comment by dluc** [comment]
@gjz22 `OpenAIChatCompletionService` methods return these types, which have a public ctor

* `ChatMessageContent`
* `StreamingChatMessageContent`
* `TextContent`
* `StreamingTextContent`

That should allow to mock SK responses.

**Comment by gjz22** [comment]
Thank you for the details! My understanding is that in order to use chat completion service function calling you must downcast `ChatMessageContent` to `OpenAIChatMessageContent`, which is a sealed class with an internal constructor making it effectively not able to be mocked.

Newer versions of semantic kernel have `FunctionCallContent`, which probably I can use to mock the output to get the function call content. In my tests I'd need to be able to mock a function / plugin + arguments for the call for my use case, which don't seem to be in the link shared above.

Thanks!
